File: WoofIOS/SwiftPackages/DesignSystem/Sources/DesignSystem/Atomic/Molecules/Button/State/ButtonState.swift
import Foundation

public enum ButtonState {
    case enabled
    case disabled
    case loading

    public func alpha(from theme: ThemeImplementing) -> Double {
        switch self {
        case .enabled: theme.alpha.enabled
        case .disabled: theme.alpha.disabled
        case .loading: theme.alpha.enabled
        }
    }

    public var isClickable: Bool {
        switch self {
        case .enabled: true
        case .disabled: false
        case .loading: false
        }
    }

    public var showsLoadingIndicator: Bool {
        switch self {
        case .enabled: false
        case .disabled: false
        case .loading: true
        }
    }
}

File: WoofAndroid/design-system/src/main/kotlin/com/perrystreet/woof/designsystem/atomic/molecules/button/state/ButtonState.kt
package com.perrystreet.woof.designsystem.atomic.molecules.button.state

import androidx.compose.runtime.Composable
import com.perrystreet.woof.designsystem.theme.Theme

enum class ButtonState {
    Enabled {
        @Composable
        override fun alpha(): Float = Theme.alpha.enabled
        override val isClickable = true
        override val showsLoadingIndicator = false
    },
    Disabled {
        @Composable
        override fun alpha(): Float = Theme.alpha.disabled
        override val isClickable = false
        override val showsLoadingIndicator = false
    },
    Loading {
        @Composable
        override fun alpha(): Float = Theme.alpha.enabled
        override val isClickable = false
        override val showsLoadingIndicator = true
    },
    ;

    @Composable
    abstract fun alpha(): Float
    abstract val isClickable: Boolean
    abstract val showsLoadingIndicator: Boolean
}
